c_src/device/glfw.c

```diff
@@ -281,80 +286,110 @@ int device_init( const device_opts_t* p_opts, device_info_t* p_info ) {
+
+  nvgBeginFrame(p_ctx, g_device_info.width, g_device_info.height, g_device_info.ratio);
+
+  // set the global transform
+  nvgTransform(p_ctx,
+               p_data->global_tx[0], p_data->global_tx[1],
+               p_data->global_tx[2], p_data->global_tx[3],
+               p_data->global_tx[4], p_data->global_tx[5]);
 }
 
-void device_end_render() {
-  glfwSwapBuffers( g_glfw_data.p_window );
+void device_begin_cursor_render(driver_data_t* p_data)
+{
+  NVGcontext* p_ctx = (NVGcontext*)p_data->v_ctx;
+  nvgTranslate(p_ctx,
+               p_data->cursor_pos[0], p_data->cursor_pos[1]);
+}
+
+void device_end_render(driver_data_t* p_data)
+{
+  NVGcontext* p_ctx = (NVGcontext*)p_data->v_ctx;
+
+  // End frame and swap front and back buffers
+  //uint64_t time = monotonic_time();
+  nvgEndFrame(p_ctx);
+  //log_info("nvgEndFrame: %" PRId64, monotonic_time() - time);
+
+  //time = monotonic_time();
+  glfwSwapBuffers(g_glfw_data.p_window);
+  //log_info("device_end_render: %" PRId64, monotonic_time() - time);
 }
 
 
 //---------------------------------------------------------
-void device_poll() {
+void device_poll()
+{
   glfwPollEvents();
 }
 
 //---------------------------------------------------------
-char* device_gl_error() {
+char* device_gl_error()
+{
   GLenum err;
   while (true)
   {
     err = glGetError();
     // check if there was a gl error
     switch (err)
     {
-      case GL_NO_ERROR:
-        return NULL;
-      case GL_INVALID_ENUM:
-        send_puts( "GL_INVALID_ENUM" );
-        break;
-      case GL_INVALID_VALUE:
-        send_puts( "GL_INVALID_VALUE" );
-        break;
-      case GL_INVALID_OPERATION:
-        send_puts( "GL_INVALID_OPERATION" );
-        break;
-      case GL_OUT_OF_MEMORY:
-        send_puts( "GL_OUT_OF_MEMORY" );
-        break;
-
+    case GL_NO_ERROR:
+      return NULL;
+    case GL_INVALID_ENUM:
+      log_error("GL_INVALID_ENUM");
+      break;
+    case GL_INVALID_VALUE:
+      log_error("GL_INVALID_VALUE");
+      break;
+    case GL_INVALID_OPERATION:
+      log_error("GL_INVALID_OPERATION");
+      break;
+    case GL_OUT_OF_MEMORY:
+      log_error("GL_OUT_OF_MEMORY");
+      break;
 #ifdef GL_STACK_UNDERFLOW
-      case GL_STACK_UNDERFLOW:
-        send_puts( "GL_STACK_UNDERFLOW" );
-        break;
+    case GL_STACK_UNDERFLOW:
+      log_error("GL_STACK_UNDERFLOW");
+      break;
 #endif
-
 #ifdef GL_STACK_OVERFLOW
-      case GL_STACK_OVERFLOW:
-        send_puts( "GL_STACK_OVERFLOW" );
-        break;
+    case GL_STACK_OVERFLOW:
+      log_error("GL_STACK_OVERFLOW");
+      break;
 #endif
-        
-      case GL_INVALID_FRAMEBUFFER_OPERATION:
-        send_puts( "GL_INVALID_FRAMEBUFFER_OPERATION" );
-        break;
-      default:
-        put_sn( "GL_OTHER:", err );
-        break;
+    case GL_INVALID_FRAMEBUFFER_OPERATION:
+      log_error("GL_INVALID_FRAMEBUFFER_OPERATION");
+      break;
+    default:
+      log_error("GL_OTHER: %d", err);
+      break;
     }
   }
 }
```

Comment:
This routine really wants to live in like a `gl_helpers.c` file.

Reply:
Moved to `c_src/device/gl_helpers.c`